lib/autoupdate/agent/process.go

```diff
@@ -0,0 +1,175 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package agent
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"log/slog"
+	"os"
+	"os/exec"
+
+	"github.com/gravitational/trace"
+)
+
+// SystemdService manages a Teleport systemd service.
+type SystemdService struct {
+	// ServiceName specifies the systemd service name.
+	ServiceName string
+	// Log contains a logger.
+	Log *slog.Logger
+}
+
+// Reload a systemd service.
+// Attempts a graceful reload before a hard restart.
+// See Process interface for more details.
+func (s SystemdService) Reload(ctx context.Context) error {
+	if err := s.checkSystem(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	// Command error codes < 0 indicate that we are unable to run the command.
+	// Errors from s.systemctl are logged along with stderr and stdout (debug only).
+
+	// If the service is not running, return ErrNotNeeded.
+	// Note systemctl reload returns an error if the unit is not active, and
+	// try-reload-or-restart is too recent of an addition for centos7.
+	code := s.systemctl(ctx, slog.LevelDebug, "is-active", "--quiet", s.ServiceName)
+	switch {
+	case code < 0:
+		return trace.Errorf("unable to determine if systemd service is active")
+	case code > 0:
+		s.Log.WarnContext(ctx, "Teleport systemd service not running.")
+		return trace.Wrap(ErrNotNeeded)
+	}
+	// Attempt graceful reload of running service.
+	code = s.systemctl(ctx, slog.LevelError, "reload", s.ServiceName)
+	switch {
+	case code < 0:
+		return trace.Errorf("unable to attempt reload of Teleport systemd service")
+	case code > 0:
+		// Graceful reload fails, try hard restart.
+		code = s.systemctl(ctx, slog.LevelError, "try-restart", s.ServiceName)
+		if code != 0 {
+			return trace.Errorf("hard restart of Teleport systemd service failed")
+		}
+		s.Log.WarnContext(ctx, "Teleport ungracefully restarted. Connections potentially dropped.")
+	default:
+		s.Log.InfoContext(ctx, "Teleport gracefully reloaded.")
+	}
+
+	// TODO(sclevine): Ensure restart was successful and verify healthcheck.
```

Comment:
@hugoShaka FYI, I didn't port the last-restart-time healthcheck from the old updater. Let me know if you think we should implement a proper healthcheck before shipping an MVP, or port the existing check over.

Reply:
👍 will port over the existing behavior in another PR